// aplx-hmw-security/src/aws-cognito/mapper.ts
import { CognitoIdentityServiceProvider } from "aws-sdk"
import { usersignup } from "./interface";
import { factory } from "../config/factory";

export class AwsCognitoMappers {
    /**
    * Creates mapper for sign up the user in aws
    * @param params 
    * @returns CognitoIdentityServiceProvider.AdminCreateUserRequest
    */
    adminCreateSignUpRequestMapper(params: usersignup): CognitoIdentityServiceProvider.AdminCreateUserRequest {
        return {
            UserPoolId: factory.getWebConfig().AWS.USER_POOL_ID,
            Username: params.email,
            DesiredDeliveryMediums: [
                'EMAIL'
            ],
            ForceAliasCreation: true,
            MessageAction: "SUPPRESS",
            TemporaryPassword: factory.getWebConfig().AWS.TEMPORARY_PASSWORD,
            UserAttributes: [
                {
                    Name: 'email',
                    Value: params.email
                },
                {
                    Name: 'phone_number',
                    Value: params.mobile
                }
            ]
        };
    }

    adminCreateSignUpResponseMapper(params: CognitoIdentityServiceProvider.AdminCreateUserResponse): JSON {
        return params as JSON;
    }

}
