// aplx-hmw-server/build/dataservice/dataservice-csrf-token.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var axios_1 = __importDefault(require("axios"));
/**
 * Holds axios framework based rest server functions to communicate with external servers
 * Headers are configured with user name and password and it will fetch ** CSRF TOKEN ** to communicate with SAP.
 * @class
 */
var DatawareServiceWithPasswordCSRFTokenAuth = /** @class */ (function () {
    function DatawareServiceWithPasswordCSRFTokenAuth(config) {
        this.username = "";
        this.password = "";
        this.headers = {
            'Authorization': "Basic " + Buffer.from(this.username + ":" + this.password).toString("base64"),
            'Content-Type': 'application/json',
            'Accept': 'application/json',
        };
        this.username = config.username;
        this.password = config.password;
    }
    DatawareServiceWithPasswordCSRFTokenAuth.prototype.setHeaders = function (results) {
        this.headers = {
            'Authorization': "Basic " + Buffer.from(this.username + ":" + this.password).toString("base64"),
            'Content-Type': 'application/json',
            'Accept': 'application/json',
        };
        if (results !== undefined) {
            // CSRF TOKEN VALIDATION. Set Cookie to header
            if (results["headers"]["set-cookie"]) {
                this.headers['Cookie'] = results["headers"]["set-cookie"];
            }
            if (results["headers"]['x-csrf-token']) {
                this.headers['x-csrf-token'] = results["headers"]["x-csrf-token"];
            }
            else {
                this.headers['x-csrf-token'] = "Fetch";
            }
        }
        else {
            this.headers['x-csrf-token'] = "Fetch";
        }
        axios_1.default.defaults.headers.common = this.headers;
        axios_1.default.defaults.responseType = 'json';
        return this.headers;
    };
    /**
     * Get web service will be called with the given url and if any headers added
     * By default it will add csrf token to the
     * @param url ip/domain in which the web service needs to communicate
     * @param headers if any headers to be appended
     */
    DatawareServiceWithPasswordCSRFTokenAuth.prototype.get = function (url, headers) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (headers) {
                axios_1.default.defaults.headers.common = _this.headers;
                axios_1.default.defaults.responseType = 'json';
            }
            else {
                _this.setHeaders();
            }
            try {
                return axios_1.default.get(url)
                    .then(function (result) {
                    return resolve(result.data);
                }).catch(function (err) {
                    return reject(err);
                });
            }
            catch (error) {
                return reject(error);
            }
        });
    };
    /**
     * Function to call post web service with Axios Framework.
     * This is Promise based function call.
     * Uses CSRF validation for OData Calls
     * @param url -> string with filters applied
     * @returns Promise Object
     */
    DatawareServiceWithPasswordCSRFTokenAuth.prototype.post = function (url, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.getCSRFToken(url).then(function (results) {
                var headers = {
                    'Authorization': "Basic " + Buffer.from(_this.username + ":" + _this.password).toString("base64"),
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                };
                if (results["headers"]["set-cookie"]) {
                    headers['Cookie'] = results["headers"]["set-cookie"];
                }
                if (results["headers"]['x-csrf-token']) {
                    headers['x-csrf-token'] = results["headers"]["x-csrf-token"];
                }
                // setHeaders(results)
                axios_1.default.post(url, data, { headers: headers }).then(function (response) {
                    return resolve(response);
                }).catch(function (err) {
                    return reject(err.response.data);
                });
            }).catch(function (err) {
                return reject(err);
            });
        });
    };
    DatawareServiceWithPasswordCSRFTokenAuth.prototype.put = function (url, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.getCSRFToken(url).then(function (results) {
                var headers = {
                    'Authorization': "Basic " + Buffer.from(_this.username + ":" + _this.password).toString("base64"),
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                };
                if (results["headers"]["set-cookie"]) {
                    headers['Cookie'] = results["headers"]["set-cookie"];
                }
                if (results["headers"]['x-csrf-token']) {
                    headers['x-csrf-token'] = results["headers"]["x-csrf-token"];
                }
                // this.setHeaders(results)
                axios_1.default.put(url, data, { headers: headers }).then(function (response) {
                    return resolve(response);
                }).catch(function (err) {
                    var errResponse = {
                        status: err.response.status,
                        statusText: err.response.statusText,
                        url: err.response.config.url,
                        data: err.response.data
                    };
                    return reject(errResponse);
                });
            }).catch(function (err) {
                return reject(err);
            });
        });
    };
    DatawareServiceWithPasswordCSRFTokenAuth.prototype.getCSRFToken = function (url) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            try {
                _this.setHeaders();
                return axios_1.default.get(url.substring(0, url.lastIndexOf("/")))
                    .then(function (result) {
                    return resolve(result);
                }).catch(function (err) {
                    return reject(err);
                });
            }
            catch (err) {
                console.log(err);
            }
        });
    };
    return DatawareServiceWithPasswordCSRFTokenAuth;
}());
exports.DatawareServiceWithPasswordCSRFTokenAuth = DatawareServiceWithPasswordCSRFTokenAuth;
