// aplx-hmw-security/build/aplx-auth/controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("../common");
// import { AplxAuthSchema } from "./schema";
var aplx_hmw_server_1 = require("aplx-hmw-server");
var AplxAuthController = /** @class */ (function () {
    function AplxAuthController() {
    }
    /**
    * Create users data for the company
    * @param params input for creating user
    * @param params.loginName email id of the user
    * @param params.password secret key of the user which will help for login
    * @param params.firstName Fisrt Name of the user
    * @param params.lastName Last Name of the user
    * @param params.stores  the user is authorized stores
    * @param params.groups the user belongs to group
    * @param param.roles User role
    * @param param.privileges
    */
    /*
     createUser(params: IUser, userToken): Promise<any> {
         return new CommonController().validateJWTToken(userToken).then((tokenData) => {
             if (!params.groups) {
                 params.groups = []
             }
             if (!params.stores) {
                 params.stores = []
             }
             if (!params.privileges) {
                 params.privileges = []
             }
             if (!params.roles) {
                 params.roles = [];
             }
             return new Promise((resolve, reject) => {
                 const userModel = AplxAuthSchema.getUserSchemaModel();
                 const user = new userModel(new AplxAuthMapper().createUserRequestMapper(params, tokenData["userId"]));
                 user.save((err, success) => {
                     if (err) {
                         return reject(err);
                     }
                     return resolve("Customer created successfully.")
                 })
             })
         });
     }
     */
    /**
    * Communicate to database and do login
    */
    /**
  * Will communicate to mongo db to check whether the user name is available there are not.
  * Based on availablity the hashed password will be validated
  * @param params - Login data
  * @param.email email address of the registered user
  * @param.password
  *
     doLogin(params: ILogin): Promise<ILoginSuccessResponse> {
        return new Promise((resolve, reject) => {
            const userModel = AplxAuthSchema.getUserSchemaModel();
            userModel.findOne({ 'loginName': params.email }, 'password status userId stores groups roles', (err, user) => {
                if (err) {
                    console.log(err);

                    return reject({
                        message: err
                    })
                }
                if (user === null) {
                    return reject({
                        message: "User not found"
                    })
                }
                if (user["status"] === UserStatus.pending || user["status"] === UserStatus.banned) {
                    if (user["status"] === UserStatus.pending) {
                        return reject({
                            message: "User is not activated. Reset the password"
                        });
                    }
                    if (user["status"] === UserStatus.banned) {
                        return reject({
                            message: "User is banned, contact administrator"
                        });
                    }
                }
                const match = new CommonController().comparePassword(params.password, user["password"]);
                if (match) {
                    const payload = {
                        userId: user["_id"],
                        stores: user["stores"],
                        roles: user["roles"],
                        groups: user["groups"],
                    }
                    const token = new CommonController().createJWTToken(payload);
                    const settings = {
                        "installationMode": "SAP/MONGODB",
                        "sap": {
                            "baseURL": "",
                            "username": "",
                            "password": "",
                            "authenticationMode": "NAMED/SINGLE",
                            "type": "S4H/S4HC/AFS/ECC/ISR/FMS"
                        },
                        "mongodb": {
                            "baseURL": "",
                            "username": "",
                            "password": ""
                        },
                        "products": {
                            "pos": ["customer", "quotation", "salesorder", "return", "quicksale", "cashdrawer", "commission", "catalog", "deliveryscheduler"],
                            "instore": ["sto", "receiving", "stockadjustment", "physicalinventory", "bumping", "movement", "inventory"]
                        },
                        "operatingCountries": ["IND", "USA"],
                        "fieldLevelHelp": false,
                        "demoMode": false,
                        "languages": [
                            {
                                "displayName": "English",
                                "dataValue": "en"
                            },
                            {
                                "displayName": "हिंदी",
                                "dataValue": "hi"
                            },
                            {
                                "displayName": "中文",
                                "dataValue": "zh"
                            }
                        ],
                        "defaultLanguage": "en",
                        "prefixCountryCode": true,
                        "phoneFormats": ["999-999-9999", "(999) 999-9999"],
                        "defaultPhoneFormat": "999-999-9999",
                        "calendars": ["12 Month Calendar", "NRF 4-4-5", "NRF 4-5-4"],
                        "defaultCalendar": "12",
                        "dateFormats": ["dd/mm/yyyy", "dd/mmm/yyyy", "dd/mm/yy", "dd/mmm/yy", "mm/dd/yyyy", "mmm/dd/yyyy", "mm/dd/yy", "mmm/dd/yy", "yyyy/mm/dd"],
                        "defaultDateFormat": "dd/mm/yyyy",
                        "numberFormats": [],
                        "defaultNumberFormat": "",
                        "nameFormats": [
                            {
                                "displayName": "First Name + Last Name",
                                "dataValue": "FL"
                            },
                            {
                                "displayName": "Last Name + First Name",
                                "dataValue": "LF"
                            }
                        ],
                        "defaultNameFormat": "FL",
                        "paymentGateway": "cayan",

                        // Point of Sale settings
                        "productsConfiguration":
                        {
                            "pos": {
                                "general": {
                                    "allowPriceOverride": false,
                                    "allowDiscounts": false,
                                    "preserveFilters": true,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 0,   // no limit
                                    "maxItems": 0,       // no limit
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "customer": {
                                    "defaultSearchRange": "7d",
                                    "integratedCamera": true,
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "quotation": {
                                    "defaultSearchRange": "7d",
                                    "managerPinForPrint": true,
                                    "preserveFilters": true,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 10,   // no limit
                                    "maxItems": 100,      // no limit
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "salesorder": {
                                    "defaultSearchRange": "7d",
                                    "allowPriceOverride": false,
                                    "allowDiscounts": false,
                                    "preserveFilters": true,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 10,   // no limit
                                    "maxItems": 100,      // no limit
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "return": {
                                    "defaultSearchRange": "7d",
                                    "returnWithoutReceipt": false,
                                    "returnMode": ["storecredit"],
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "quicksale": {
                                    "allowPriceOverride": false,
                                    "priceOverrideToleranceType": "VALUE/PERCENT",
                                    "priceOverrideToleranceValue": 0.00,
                                    "allowDiscounts": false,
                                    "modeOfPayments": ["cash", "card", "cheque", "account", "greensky"],
                                    "allowSplitPayments": false,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 10,   // no limit
                                    "maxItems": 100      // no limit
                                },
                                "cashdrawer": {
                                    "dayBeginMandatory": true,
                                    "dayEndMandatory": true,
                                    "allCountersClosedForDayEnd": true,
                                    "allowFutureDateDayBegin": true,
                                    "openAllCountersOnDayBegin": false,
                                    "closeCounterToleranceType": "VALUE/PERCENT",
                                    "closeCounterToleranceValue": 0.00,
                                    "dayEndToleranceType": "VALUE/PERCENT",
                                    "dayEndToleranceValue": 0.00,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "commission": {
                                    "defaultSearchRange": "7d",
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "catalog": {
                                    "currentStore": true
                                },
                                "deliveryscheduler": {
                                    "defaultSearchRange": "2d",
                                    "preserveFilters": true
                                }
                            },
                            "instore": {
                                "general": {
                                    "preserveFilters": true
                                },
                                "sto": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 0,   // no limit
                                    "maxItems": 0,       // no limit
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "receiving": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "stockadjustment": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "minOrderQty": 1,
                                    "maxOrderQty": 0,   // no limit
                                    "maxItems": 0,       // no limit
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "physicalinventory": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "bumping": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "movement": {
                                    "defaultSearchRange": "7d",
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                },
                                "inventory": {
                                    "currentStore": true,
                                    "preserveFilters": true,
                                    "listOptions": {
                                        "rowsPerPage": 10,
                                        "rowsOptions": [5, 10, 15, 20, 25],
                                        "enableSorting": true,
                                        "enableRowSelection": false,
                                        "paginator": true,
                                        "pageLinks": 5
                                    }
                                }
                            }
                        },
                        "currencyOptions": {
                            "showSymbol": false,
                            "format": "1.2-2"
                        },
                        "quantityOptions": {
                            "format": "1.2-2",
                            "mask": "999999.99"
                        },
                        "searchDateRangeDays": 600,
                        "uploads": {
                            "fileType": "",
                            "fileSize": ""
                        }
                    };
                    const privileges = ['*'];
                    const roles = ['admin'];

                    return resolve({
                        auth_token: token,
                        message: "Login is successfull",
                        settings: settings,
                        privileges: privileges,
                        roles: roles
                    });
                }
                else {
                    return reject({
                        message: "Password incorrect"
                    })
                }
            })
        })
     }; */
    AplxAuthController.prototype.doLogin = function (params) {
        return new Promise(function (resolve, reject) {
            // Get encrypted password version for the given password
            var encPassword = new common_1.CommonController().createHashPassword(params.password);
            try {
                aplx_hmw_server_1.DB.query("select core.do_login ('" + params.email + "', '" + encPassword + "')", null).then(function (row) {
                    if (row[0].do_login.status == '200') {
                        var user = row[0].do_login;
                        var payload = {
                            userId: user.user_info["user_id"],
                        };
                        var token = new common_1.CommonController().createJWTToken(payload);
                        var settings = {
                            "installationMode": "SAP/MONGODB",
                            "sap": {
                                "baseURL": "",
                                "username": "",
                                "password": "",
                                "authenticationMode": "NAMED/SINGLE",
                                "type": "S4H/S4HC/AFS/ECC/ISR/FMS"
                            },
                            "mongodb": {
                                "baseURL": "",
                                "username": "",
                                "password": ""
                            },
                            "products": {
                                "pos": ["customer", "quotation", "salesorder", "return", "quicksale", "cashdrawer", "commission", "catalog", "deliveryscheduler"],
                                "instore": ["sto", "receiving", "stockadjustment", "physicalinventory", "bumping", "movement", "inventory"]
                            },
                            "operatingCountries": ["IND", "USA"],
                            "fieldLevelHelp": false,
                            "demoMode": false,
                            "languages": [
                                {
                                    "displayName": "English",
                                    "dataValue": "en"
                                },
                                {
                                    "displayName": "हिंदी",
                                    "dataValue": "hi"
                                },
                                {
                                    "displayName": "中文",
                                    "dataValue": "zh"
                                }
                            ],
                            "defaultLanguage": "en",
                            "prefixCountryCode": true,
                            "phoneFormats": ["999-999-9999", "(999) 999-9999"],
                            "defaultPhoneFormat": "999-999-9999",
                            "calendars": ["12 Month Calendar", "NRF 4-4-5", "NRF 4-5-4"],
                            "defaultCalendar": "12",
                            "dateFormats": ["dd/mm/yyyy", "dd/mmm/yyyy", "dd/mm/yy", "dd/mmm/yy", "mm/dd/yyyy", "mmm/dd/yyyy", "mm/dd/yy", "mmm/dd/yy", "yyyy/mm/dd"],
                            "defaultDateFormat": "dd/mm/yyyy",
                            "numberFormats": [],
                            "defaultNumberFormat": "",
                            "nameFormats": [
                                {
                                    "displayName": "First Name + Last Name",
                                    "dataValue": "FL"
                                },
                                {
                                    "displayName": "Last Name + First Name",
                                    "dataValue": "LF"
                                }
                            ],
                            "defaultNameFormat": "FL",
                            "paymentGateway": "cayan",
                            // Point of Sale settings
                            "productsConfiguration": {
                                "pos": {
                                    "general": {
                                        "allowPriceOverride": false,
                                        "allowDiscounts": false,
                                        "preserveFilters": true,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 0,
                                        "maxItems": 0,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "customer": {
                                        "defaultSearchRange": "7d",
                                        "integratedCamera": true,
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "quotation": {
                                        "defaultSearchRange": "7d",
                                        "managerPinForPrint": true,
                                        "preserveFilters": true,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 10,
                                        "maxItems": 100,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "salesorder": {
                                        "defaultSearchRange": "7d",
                                        "allowPriceOverride": false,
                                        "allowDiscounts": false,
                                        "preserveFilters": true,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 10,
                                        "maxItems": 100,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "return": {
                                        "defaultSearchRange": "7d",
                                        "returnWithoutReceipt": false,
                                        "returnMode": ["storecredit"],
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "quicksale": {
                                        "allowPriceOverride": false,
                                        "priceOverrideToleranceType": "VALUE/PERCENT",
                                        "priceOverrideToleranceValue": 0.00,
                                        "allowDiscounts": false,
                                        "modeOfPayments": ["cash", "card", "cheque", "account", "greensky"],
                                        "allowSplitPayments": false,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 10,
                                        "maxItems": 100 // no limit
                                    },
                                    "cashdrawer": {
                                        "dayBeginMandatory": true,
                                        "dayEndMandatory": true,
                                        "allCountersClosedForDayEnd": true,
                                        "allowFutureDateDayBegin": true,
                                        "openAllCountersOnDayBegin": false,
                                        "closeCounterToleranceType": "VALUE/PERCENT",
                                        "closeCounterToleranceValue": 0.00,
                                        "dayEndToleranceType": "VALUE/PERCENT",
                                        "dayEndToleranceValue": 0.00,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "commission": {
                                        "defaultSearchRange": "7d",
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "catalog": {
                                        "currentStore": true
                                    },
                                    "deliveryscheduler": {
                                        "defaultSearchRange": "2d",
                                        "preserveFilters": true
                                    }
                                },
                                "instore": {
                                    "general": {
                                        "preserveFilters": true
                                    },
                                    "sto": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 0,
                                        "maxItems": 0,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "receiving": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "stockadjustment": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "minOrderQty": 1,
                                        "maxOrderQty": 0,
                                        "maxItems": 0,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "physicalinventory": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "bumping": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "movement": {
                                        "defaultSearchRange": "7d",
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    },
                                    "inventory": {
                                        "currentStore": true,
                                        "preserveFilters": true,
                                        "listOptions": {
                                            "rowsPerPage": 10,
                                            "rowsOptions": [5, 10, 15, 20, 25],
                                            "enableSorting": true,
                                            "enableRowSelection": false,
                                            "paginator": true,
                                            "pageLinks": 5
                                        }
                                    }
                                }
                            },
                            "currencyOptions": {
                                "showSymbol": false,
                                "format": "1.2-2"
                            },
                            "quantityOptions": {
                                "format": "1.2-2",
                                "mask": "999999.99"
                            },
                            "searchDateRangeDays": 600,
                            "uploads": {
                                "fileType": "",
                                "fileSize": ""
                            }
                        };
                        return resolve({
                            auth_token: token,
                            message: "Login is successfull",
                            userInfo: user.user_info,
                            settings: settings,
                            privileges: user.urls,
                            roles: user.role_info,
                            storeInfo: user.storeInfo,
                            counterInfo: user.counters
                        });
                    }
                    else {
                        return reject("Invalid Credentials");
                    }
                });
            }
            catch (error) {
                return reject(error);
            }
        });
    };
    return AplxAuthController;
}());
exports.AplxAuthController = AplxAuthController;
