// aplx-hmw-security/build/aplx-auth/mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var interface_1 = require("../common/interface");
var AplxAuthMapper = /** @class */ (function () {
    function AplxAuthMapper() {
    }
    /**
     * Will map the response of the user which to be created in aplx database
     * @param params refer IUser
     * @param createdBy user id or unique field to identify which user created the user
     */
    AplxAuthMapper.prototype.createUserRequestMapper = function (params, createdBy) {
        return {
            password: params.password,
            loginName: params.loginName,
            firstName: params.firstName,
            lastName: params.lastName,
            status: interface_1.UserStatus.active,
            stores: params.stores,
            groups: params.groups,
            roles: params.roles,
            privileges: params.privileges,
            createdBy: createdBy
        };
    };
    /**
     *
     */
    AplxAuthMapper.prototype.loginResponseMapper = function (data) {
        return {
            message: "Login success",
            token: data.token
        };
    };
    return AplxAuthMapper;
}());
exports.AplxAuthMapper = AplxAuthMapper;
