// aplx-hmw-security/build/aws-cognito/mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var factory_1 = require("../config/factory");
var AwsCognitoMappers = /** @class */ (function () {
    function AwsCognitoMappers() {
    }
    /**
    * Creates mapper for sign up the user in aws
    * @param params
    * @returns CognitoIdentityServiceProvider.AdminCreateUserRequest
    */
    AwsCognitoMappers.prototype.adminCreateSignUpRequestMapper = function (params) {
        return {
            UserPoolId: factory_1.factory.getWebConfig().AWS.USER_POOL_ID,
            Username: params.email,
            DesiredDeliveryMediums: [
                'EMAIL'
            ],
            ForceAliasCreation: true,
            MessageAction: "SUPPRESS",
            TemporaryPassword: factory_1.factory.getWebConfig().AWS.TEMPORARY_PASSWORD,
            UserAttributes: [
                {
                    Name: 'email',
                    Value: params.email
                },
                {
                    Name: 'phone_number',
                    Value: params.mobile
                }
            ]
        };
    };
    AwsCognitoMappers.prototype.adminCreateSignUpResponseMapper = function (params) {
        return params;
    };
    return AwsCognitoMappers;
}());
exports.AwsCognitoMappers = AwsCognitoMappers;
